File: shipbob_to_gorgias/kotlin/app/src/main/kotlin/sb2gorgias/Http.kt
package sb2gorgias

import io.github.oshai.kotlinlogging.KotlinLogging
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Instant
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import kotlin.jvm.optionals.getOrNull
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

private val logger = KotlinLogging.logger {}

/** Statuses worth retrying: rate limiting, plus the gateway errors both APIs return under load. */
private val RETRY_STATUSES = setOf(429, 502, 503, 504)

/**
 * Total attempts, the first included. Pandium does not retry a failed run, so a transient
 * 429 has to be absorbed here or the whole run is lost.
 */
private const val MAX_ATTEMPTS = 6

/** The longest a `Retry-After` is honoured for. */
private val MAX_RETRY_AFTER = 60.seconds

/**
 * How long [response] asked the client to wait, or `null` if it did not ask or sent a
 * header this does not understand — either way the doubling backoff is a safe fallback.
 * RFC 9110 allows both a number of seconds and an HTTP date.
 */
private fun retryAfter(response: HttpResponse<*>): Duration? {
    val header = response.headers().firstValue("retry-after").getOrNull()?.trim() ?: return null
    val seconds =
        header.toLongOrNull()
            ?: runCatching {
                ZonedDateTime.parse(header, DateTimeFormatter.RFC_1123_DATE_TIME).toEpochSecond() -
                    Instant.now().epochSecond
            }.getOrNull()
            ?: return null
    // A date that has already passed reads as a negative wait, which means "now".
    return seconds.coerceIn(0, MAX_RETRY_AFTER.inWholeSeconds).seconds
}

/**
 * A very small JSON-over-HTTP client, shared by both API clients.
 *
 * Both APIs speak bearer-authenticated JSON and both rate-limit, so all this adds over the
 * JDK's [HttpClient] are the standing headers and a bounded retry. Blocking on purpose: the
 * run is sequential from start to finish.
 */
class ApiClient(
    private val baseUrl: String,
    private val authorization: String,
    /** How long to wait before the first retry; doubled before each one after that. */
    private val backoff: Duration,
) {
    private val http: HttpClient = HttpClient.newHttpClient()

    fun get(path: String, vararg query: Pair<String, String>): JsonElement =
        send(request(path, query.toList()).GET().build())

    fun post(path: String, body: JsonElement): JsonElement =
        send(request(path).POST(HttpRequest.BodyPublishers.ofString(body.toString())).build())

    fun put(path: String, body: JsonElement): JsonElement =
        send(request(path).PUT(HttpRequest.BodyPublishers.ofString(body.toString())).build())

    private fun request(path: String, query: List<Pair<String, String>> = emptyList()): HttpRequest.Builder =
        HttpRequest.newBuilder(URI.create(baseUrl + path + queryString(query)))
            .header("accept", "application/json")
            .header("content-type", "application/json")
            .header("authorization", authorization)

    private fun queryString(query: List<Pair<String, String>>): String =
        if (query.isEmpty()) {
            ""
        } else {
            query.joinToString("&", prefix = "?") { (key, value) ->
                "$key=${URLEncoder.encode(value, Charsets.UTF_8)}"
            }
        }

    /**
     * Send [request] until it answers with something other than a retryable status, then
     * parse the body as JSON.
     *
     * A non-2xx that survives the retries throws, carrying the status and the body. Nothing
     * here maps a failure onto an empty result: the cron flow reads an empty page as "the
     * query is exhausted, commit the cursor", so the two have to stay distinct.
     */
    private fun send(request: HttpRequest): JsonElement {
        var wait = backoff
        repeat(MAX_ATTEMPTS) { attempt ->
            val response = http.send(request, HttpResponse.BodyHandlers.ofString())
            val status = response.statusCode()

            if (status in RETRY_STATUSES && attempt < MAX_ATTEMPTS - 1) {
                val pause = retryAfter(response) ?: wait
                logger.warn { "HTTP $status from ${request.uri()}; retrying in $pause (attempt ${attempt + 1})" }
                Thread.sleep(pause.inWholeMilliseconds)
                wait *= 2
                return@repeat
            }

            val body = response.body().orEmpty()
            check(status in 200..299) { "HTTP $status from ${request.uri()}: $body" }
            // A 204, and some 200s, come back empty; reading that as JSON null lets
            // callers index into the result either way.
            return if (body.isBlank()) JsonNull else json.parseToJsonElement(body)
        }
        error("the retry loop returns or retries on every attempt")
    }
}

File: SHIPBOB_TO_GORGIAS/go/http.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// retryStatuses are the response codes worth retrying — matches every other
// language's port here (429, 502, 503, 504).
var retryStatuses = map[int]bool{429: true, 502: true, 503: true, 504: true}

// retryClient is a small HTTP client with exponential backoff retry, scoped to one
// API's base URL and auth header. doRequest is injectable so tests can stub the
// network without a full http.RoundTripper mock.
type retryClient struct {
	doRequest     func(*http.Request) (*http.Response, error)
	baseURL       string
	authorization string
	backoff       time.Duration
	retryMethods  map[string]bool
	maxAttempts   int
}

func newRetryClient(baseURL, authorization string, backoff time.Duration, retryMethods []string) *retryClient {
	methods := make(map[string]bool, len(retryMethods))
	for _, m := range retryMethods {
		methods[m] = true
	}
	httpClient := &http.Client{Timeout: 30 * time.Second}
	return &retryClient{
		doRequest:     httpClient.Do,
		baseURL:       baseURL,
		authorization: authorization,
		backoff:       backoff,
		retryMethods:  methods,
		maxAttempts:   6,
	}
}

func (c *retryClient) get(path string, query url.Values) (any, error) {
	full := c.baseURL + path
	if len(query) > 0 {
		full += "?" + query.Encode()
	}
	return c.send(http.MethodGet, full, nil)
}

func (c *retryClient) post(path string, body any) (any, error) {
	return c.send(http.MethodPost, c.baseURL+path, body)
}

func (c *retryClient) put(path string, body any) (any, error) {
	return c.send(http.MethodPut, c.baseURL+path, body)
}

// send performs one request, retrying on a network error or a retryable status
// code when the method is allowed to retry. Backoff is backoff * 2^(attempt-1)
// for attempts after the first, no jitter.
//
// A response with an empty body decodes to (nil, nil), not an error — that's what
// lets a caller distinguish "genuinely nothing here" from a real failure.
func (c *retryClient) send(method, fullURL string, body any) (any, error) {
	var bodyBytes []byte
	if body != nil {
		var err error
		bodyBytes, err = json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("could not encode request body: %w", err)
		}
	}

	var lastErr error
	for attempt := 1; attempt <= c.maxAttempts; attempt++ {
		if attempt > 1 {
			time.Sleep(c.backoff * time.Duration(1<<(attempt-2)))
		}

		req, err := http.NewRequest(method, fullURL, bytes.NewReader(bodyBytes))
		if err != nil {
			return nil, fmt.Errorf("could not build request: %w", err)
		}
		req.Header.Set("accept", "application/json")
		req.Header.Set("content-type", "application/json")
		req.Header.Set("Authorization", c.authorization)

		res, err := c.doRequest(req)
		if err != nil {
			lastErr = err
			if c.retryMethods[method] {
				continue
			}
			return nil, err
		}

		respBody, readErr := io.ReadAll(res.Body)
		res.Body.Close()
		if readErr != nil {
			return nil, fmt.Errorf("could not read response body: %w", readErr)
		}

		if res.StatusCode >= 400 {
			lastErr = fmt.Errorf("HTTP %d: %s", res.StatusCode, respBody)
			if retryStatuses[res.StatusCode] && c.retryMethods[method] {
				continue
			}
			return nil, lastErr
		}

		if len(bytes.TrimSpace(respBody)) == 0 {
			return nil, nil
		}
		var data any
		if err := json.Unmarshal(respBody, &data); err != nil {
			return nil, fmt.Errorf("could not decode response body: %w", err)
		}
		return data, nil
	}
	return nil, lastErr
}
